src/WebpackLicensePlugin.ts

```diff
@@ -69,7 +71,14 @@ export default class WebpackLicensePlugin implements IWebpackPlugin {
     alertAggregator.flushAlerts(pluginName)
 
     const chunkIterator = new WebpackChunkIterator()
-    const filenames = chunkIterator.iterateChunks(compilation, chunks)
+    for (const filename of chunkIterator.iterateChunks(compilation, chunks)) {
+      this.filenames.add(filename)
+    }
+
+    if (compilation.compiler?.isChild()) {
```

Comment:
Just asking to understand this better. With multiple (child) compilers, `handleChunkAssetOptimization` is called once for each compiler. When the compiler is a child compiler, the method is exited prematurely after identifying related filenames and adding them to the set, but before doing any other meaningful work.

Because `handleChunkAssetOptimization` is (by design?) always called last with the root compiler, on the last invocation with that root compiler the identified files from that compilation are added to all of the files previously added by all the child compilers and then the "normal" rest of the flow is performed.

Are we sure that, under all circumstances, an invocation of `handleChunkAssetOptimization` with a child compiler can not happen after it was already invoked with the root compiler?

Reply:
The childCompilers are created by other plugins using the [`createChildCompiler`](https://webpack.js.org/api/compilation-object/#createchildcompiler) function. 

> It copies all hooks and plugins from the parent (or top-level compiler)

That is the reason why webpack-license-plugin will also run in the child compilers. I guess at what point a child compiler runs depends on the hook a plugin used to create and run the child compiler. Assuming we have he following hook order:
- hookBefore (an imaginary name)
- optimizeChunkAssets
- hookAfter (also imaginary)

If a plugin runs its child compiler in `hookBefore`, webpack-license-plugin's `handleChunkAssetOptimization` function should be called with the child compiler before it is called with the root compiler (because the root compiler has not reached the `optimizeChunkAssets` yet).

If a plugin runs its child compiler in the `optimizeChunkAssets` I think it depends on the plugin order inside the webpack config whether the child compiler is called first or the webpack-license-plugin's `handleChunkAssetOptimization`.

If a plugin runs its child compiler in the `hookAfter`, webpack-license-plugin's `handleChunkAssetOptimization` function should be called with the root compiler first, then with the child compiler. In this case the license information from the child compiler gets lost. On the other hand, without the changes in this PR it will probably simply crash because there would be two assets with the same name again.

It might be worth trying to seperating the dependency collecting and the file creation even more. The dependency collecting could stay in the `optimizeChunkAssets` hook but the file creation could maybe move into the `emit` compiler hook. I think I will try that next.

Reply:
Creating the asset in the `emit` hook technically works but webpack 5 tells me that I should not do that.
```
(node:6965) [DEP_WEBPACK_COMPILATION_ASSETS] DeprecationWarning: Compilation.assets will be frozen in future, all modifications are deprecated.
BREAKING CHANGE: No more changes should happen to Compilation.assets after sealing the Compilation.
        Do changes to assets earlier, e. g. in Compilation.hooks.processAssets.
        Make sure to select an appropriate stage from Compilation.PROCESS_ASSETS_STAGE_*.
```

I went and logged all compilation hooks that get triggered in my reproduction repo:
```
root: addEntry
root: childCompiler
HtmlWebpackCompiler: addEntry
HtmlWebpackCompiler: addEntry
HtmlWebpackCompiler: buildModule
HtmlWebpackCompiler: succeedModule
HtmlWebpackCompiler: succeedEntry
root: buildModule
HtmlWebpackCompiler: buildModule
root: succeedModule
HtmlWebpackCompiler: succeedModule
HtmlWebpackCompiler: succeedEntry
HtmlWebpackCompiler: finishModules
HtmlWebpackCompiler: seal
HtmlWebpackCompiler: optimizeDependencies
HtmlWebpackCompiler: afterOptimizeDependencies
HtmlWebpackCompiler: beforeChunks
HtmlWebpackCompiler: afterChunks
HtmlWebpackCompiler: optimize
HtmlWebpackCompiler: optimizeModules
HtmlWebpackCompiler: afterOptimizeModules
HtmlWebpackCompiler: optimizeChunks
HtmlWebpackCompiler: afterOptimizeChunks
HtmlWebpackCompiler: optimizeTree
HtmlWebpackCompiler: afterOptimizeTree
HtmlWebpackCompiler: optimizeChunkModules
HtmlWebpackCompiler: afterOptimizeChunkModules
HtmlWebpackCompiler: shouldRecord
HtmlWebpackCompiler: reviveModules
HtmlWebpackCompiler: beforeModuleIds
HtmlWebpackCompiler: moduleIds
HtmlWebpackCompiler: optimizeModuleIds
HtmlWebpackCompiler: afterOptimizeModuleIds
HtmlWebpackCompiler: reviveChunks
HtmlWebpackCompiler: beforeChunkIds
HtmlWebpackCompiler: chunkIds
HtmlWebpackCompiler: optimizeChunkIds
HtmlWebpackCompiler: afterOptimizeChunkIds
HtmlWebpackCompiler: recordModules
HtmlWebpackCompiler: recordChunks
HtmlWebpackCompiler: optimizeCodeGeneration
HtmlWebpackCompiler: beforeModuleHash
HtmlWebpackCompiler: afterModuleHash
HtmlWebpackCompiler: beforeCodeGeneration
HtmlWebpackCompiler: afterCodeGeneration
HtmlWebpackCompiler: beforeRuntimeRequirements
HtmlWebpackCompiler: additionalModuleRuntimeRequirements
HtmlWebpackCompiler: additionalModuleRuntimeRequirements
HtmlWebpackCompiler: additionalChunkRuntimeRequirements
HtmlWebpackCompiler: additionalTreeRuntimeRequirements
HtmlWebpackCompiler: runtimeModule
HtmlWebpackCompiler: runtimeModule
HtmlWebpackCompiler: runtimeModule
HtmlWebpackCompiler: runtimeModule
HtmlWebpackCompiler: afterRuntimeRequirements
HtmlWebpackCompiler: beforeHash
HtmlWebpackCompiler: assetPath
HtmlWebpackCompiler: assetPath
HtmlWebpackCompiler: chunkHash
HtmlWebpackCompiler: assetPath
HtmlWebpackCompiler: contentHash
HtmlWebpackCompiler: assetPath
HtmlWebpackCompiler: fullHash
HtmlWebpackCompiler: afterHash
HtmlWebpackCompiler: recordHash
HtmlWebpackCompiler: beforeModuleAssets
HtmlWebpackCompiler: shouldGenerateChunkAssets
HtmlWebpackCompiler: beforeChunkAssets
HtmlWebpackCompiler: renderManifest
HtmlWebpackCompiler: assetPath
HtmlWebpackCompiler: assetPath
HtmlWebpackCompiler: assetPath
HtmlWebpackCompiler: chunkAsset
HtmlWebpackCompiler: additionalChunkAssets
HtmlWebpackCompiler: additionalAssets
HtmlWebpackCompiler: optimizeAssets
HtmlWebpackCompiler: processAssets
HtmlWebpackCompiler: optimizeChunkAssets
HtmlWebpackCompiler: afterOptimizeChunkAssets
HtmlWebpackCompiler: afterOptimizeAssets
HtmlWebpackCompiler: afterProcessAssets
HtmlWebpackCompiler: record
HtmlWebpackCompiler: needAdditionalSeal
HtmlWebpackCompiler: afterSeal
root: buildModule
root: childCompiler
mini-css-extract-plugin: addEntry
mini-css-extract-plugin: buildModule
mini-css-extract-plugin: succeedModule
mini-css-extract-plugin: buildModule
mini-css-extract-plugin: succeedModule
mini-css-extract-plugin: succeedEntry
mini-css-extract-plugin: finishModules
mini-css-extract-plugin: seal
mini-css-extract-plugin: optimizeDependencies
mini-css-extract-plugin: afterOptimizeDependencies
mini-css-extract-plugin: beforeChunks
mini-css-extract-plugin: afterChunks
mini-css-extract-plugin: optimize
mini-css-extract-plugin: optimizeModules
mini-css-extract-plugin: afterOptimizeModules
mini-css-extract-plugin: optimizeChunks
mini-css-extract-plugin: afterOptimizeChunks
mini-css-extract-plugin: optimizeTree
mini-css-extract-plugin: afterOptimizeTree
mini-css-extract-plugin: optimizeChunkModules
mini-css-extract-plugin: afterOptimizeChunkModules
mini-css-extract-plugin: shouldRecord
mini-css-extract-plugin: reviveModules
mini-css-extract-plugin: beforeModuleIds
mini-css-extract-plugin: moduleIds
mini-css-extract-plugin: optimizeModuleIds
mini-css-extract-plugin: afterOptimizeModuleIds
mini-css-extract-plugin: reviveChunks
mini-css-extract-plugin: beforeChunkIds
mini-css-extract-plugin: chunkIds
mini-css-extract-plugin: optimizeChunkIds
mini-css-extract-plugin: afterOptimizeChunkIds
mini-css-extract-plugin: recordModules
mini-css-extract-plugin: recordChunks
mini-css-extract-plugin: optimizeCodeGeneration
mini-css-extract-plugin: beforeModuleHash
mini-css-extract-plugin: afterModuleHash
mini-css-extract-plugin: beforeCodeGeneration
mini-css-extract-plugin: afterCodeGeneration
mini-css-extract-plugin: beforeRuntimeRequirements
mini-css-extract-plugin: additionalModuleRuntimeRequirements
mini-css-extract-plugin: additionalModuleRuntimeRequirements
mini-css-extract-plugin: additionalChunkRuntimeRequirements
mini-css-extract-plugin: additionalTreeRuntimeRequirements
mini-css-extract-plugin: runtimeModule
mini-css-extract-plugin: runtimeModule
mini-css-extract-plugin: runtimeModule
mini-css-extract-plugin: runtimeModule
mini-css-extract-plugin: afterRuntimeRequirements
mini-css-extract-plugin: beforeHash
mini-css-extract-plugin: chunkHash
mini-css-extract-plugin: assetPath
mini-css-extract-plugin: assetPath
mini-css-extract-plugin: contentHash
mini-css-extract-plugin: assetPath
mini-css-extract-plugin: assetPath
mini-css-extract-plugin: fullHash
mini-css-extract-plugin: afterHash
mini-css-extract-plugin: recordHash
mini-css-extract-plugin: beforeModuleAssets
mini-css-extract-plugin: shouldGenerateChunkAssets
mini-css-extract-plugin: beforeChunkAssets
mini-css-extract-plugin: renderManifest
mini-css-extract-plugin: assetPath
mini-css-extract-plugin: assetPath
mini-css-extract-plugin: assetPath
mini-css-extract-plugin: assetPath
mini-css-extract-plugin: assetPath
mini-css-extract-plugin: chunkAsset
mini-css-extract-plugin: additionalChunkAssets
mini-css-extract-plugin: additionalAssets
mini-css-extract-plugin: optimizeAssets
mini-css-extract-plugin: processAssets
mini-css-extract-plugin: optimizeChunkAssets
mini-css-extract-plugin: afterOptimizeChunkAssets
mini-css-extract-plugin: afterOptimizeAssets
mini-css-extract-plugin: afterProcessAssets
mini-css-extract-plugin: record
mini-css-extract-plugin: needAdditionalSeal
mini-css-extract-plugin: afterSeal
root: succeedModule
root: buildModule
root: succeedModule
root: succeedEntry
root: finishModules
root: seal
root: optimizeDependencies
root: afterOptimizeDependencies
root: beforeChunks
root: afterChunks
root: optimize
root: optimizeModules
root: afterOptimizeModules
root: optimizeChunks
root: afterOptimizeChunks
root: optimizeTree
root: afterOptimizeTree
root: optimizeChunkModules
root: afterOptimizeChunkModules
root: shouldRecord
root: reviveModules
root: beforeModuleIds
root: moduleIds
root: optimizeModuleIds
root: afterOptimizeModuleIds
root: reviveChunks
root: beforeChunkIds
root: chunkIds
root: optimizeChunkIds
root: afterOptimizeChunkIds
root: recordModules
root: recordChunks
root: optimizeCodeGeneration
root: beforeModuleHash
root: afterModuleHash
root: beforeCodeGeneration
root: afterCodeGeneration
root: beforeRuntimeRequirements
root: additionalModuleRuntimeRequirements
root: additionalModuleRuntimeRequirements
root: additionalModuleRuntimeRequirements
root: additionalChunkRuntimeRequirements
root: additionalTreeRuntimeRequirements
root: runtimeModule
root: afterRuntimeRequirements
root: beforeHash
root: chunkHash
root: contentHash
root: fullHash
root: afterHash
root: recordHash
root: beforeModuleAssets
root: shouldGenerateChunkAssets
root: beforeChunkAssets
root: renderManifest
root: assetPath
root: assetPath
root: chunkAsset
root: assetPath
root: chunkAsset
root: additionalChunkAssets
root: additionalAssets
root: optimizeAssets
root: processAssets
root: processAdditionalAssets
root: optimizeChunkAssets
root: afterOptimizeChunkAssets
root: assetPath
root: processAdditionalAssets
root: afterOptimizeAssets
root: afterProcessAssets
root: record
root: needAdditionalSeal
root: afterSeal
root: assetPath
root: needAdditionalPass
```
Seems to me that `optimizeChunkAssets` is already the last hook that can create/edit assets.